app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt

```diff
@@ -1,33 +1,250 @@
+            when (locationType) {
+                VariableLocationType.MEMORY -> {
+                    variablesStackOffsets[variable] = variablesTotalOffset
+                    variablesTotalOffset += memoryUnitSize
+                }
+
+                VariableLocationType.REGISTER -> {
+                    variablesRegisters[variable] = createRegisterFor(variable)
+                }
+            }
+        }
+    }
+
+    override fun genCall(
+        args: List<IntermediateFormTreeNode>,
+    ): FunctionDetailsGeneratorInterface.FunctionCallIntermediateForm {
+        // First, it moves arguments to appropriate registers (or pushes to stack) according to x86 call convention.
+        // Then, it adds call instruction to actually call a given function
+        // At the end it creates IFTNode to get function result
+
+        val cfgBuilder = ControlFlowGraphBuilder()
+
+        for ((arg, register) in args zip argPositionToRegister) {
+            val node = IntermediateFormTreeNode.RegisterWrite(register, arg)
+            cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, node)
+        }
+        for (arg in args.drop(argPositionToRegister.size).reversed()) {
+            val node = IntermediateFormTreeNode.StackPush(arg)
+            cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, node)
+        }
+
+        cfgBuilder.addLinkFromAllFinalRoots(
+            CFGLinkType.UNCONDITIONAL,
+            IntermediateFormTreeNode.Call(functionLocationInCode)
+        )
+
+        var readResultNode: IFTNode? = null
+        if (variableToStoreFunctionResult != null) {
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                IntermediateFormTreeNode.RegisterWrite(Register.RAX, genRead(variableToStoreFunctionResult, true))
+            )
+            readResultNode = IntermediateFormTreeNode.RegisterRead(Register.RAX)
+        }
+        return FunctionDetailsGeneratorInterface.FunctionCallIntermediateForm(cfgBuilder.build(), readResultNode)
     }
 
     override fun genPrologue(): ControlFlowGraph {
-        return TODO()
+        val cfgBuilder = ControlFlowGraphBuilder()
+
+        // move args from Registers and Stack
+        for ((param, register) in parameters zip argPositionToRegister) {
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                genWrite(
+                    param,
+                    IntermediateFormTreeNode.RegisterRead(register),
+                    true
+                )
+            )
+        }
+        for (param in parameters.drop(argPositionToRegister.size)) {
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                genWrite(
+                    param,
+                    IntermediateFormTreeNode.StackPop(),
+                    true
+                )
+            )
+        }
+
+        // backup callee-saved registers
```

Comment:
It is tricky, because we'd have to watch out to not assign not already backed up callee-saved register to that virtual register storing the copy

Reply:
Moved pushing them to stack explicitly after allocating space on stack for local variables, but if you think my previous comment isn't a problem, I'd be happy to employ your solution

Reply:
Well, we're going to have some method of forbidding arbitrary register assignments (by adding extra edges to permanently colored vertices). We don't know yet how such an information will be passed there, so maybe (probably) some interfaces will have to be updated, but I'm pretty sure that it will be possible to do it in a clear way.